File: rust/examples/transpose.rs
#![no_std]
#![no_main]
#![allow(unused)]

// Crates being used
// =================

use riscv_rt::entry;

extern crate nocl;
use nocl::*;
use nocl::rand::*;
use nocl::prims::*;

extern crate alloc;
use alloc::vec;
use alloc::vec::*;
use alloc::boxed::*;

// Benchmark
// =========

const SQUARE_SIZE: usize = prims::config::SIMT_LANES as usize;

struct Transpose {
  width  : usize,
  height : usize,
  input  : Box<[i32]>,
  output : Box<[i32]>
}

impl Code for Transpose {

#[inline(always)]
fn run (my : &My, shared : &mut Scratch, params: &mut Transpose) {
  let mut square = alloc::<i32>(shared, (SQUARE_SIZE+1) * SQUARE_SIZE);

  // Origin of square within matrix
  let origin_x = my.block_idx.x * my.block_dim.x;
  let origin_y = my.block_idx.y * my.block_dim.x;

  // Load square
  for y in (my.thread_idx.y .. my.block_dim.x).step_by(my.block_dim.y) {
    square[y*(SQUARE_SIZE+1) + my.thread_idx.x] =
      params.input[(origin_y + y)*params.width + origin_x + my.thread_idx.x]
  }

  syncthreads();
    
  // Store square
  for y in (my.thread_idx.y .. my.block_dim.x).step_by(my.block_dim.y) {
    params.output[(origin_x + y)*params.height + origin_y + my.thread_idx.x] =
      square[my.thread_idx.x*(SQUARE_SIZE+1) + y]
  }
}

}

#[entry]
fn main() -> ! {
  nocl_init();

  // Matrix size for benchmarking
  #[cfg(not(feature = "large_data_set"))]
  const WIDTH : usize = 256;
  #[cfg(not(feature = "large_data_set"))]
  const HEIGHT : usize = 64;
  #[cfg(feature = "large_data_set")]
  const WIDTH : usize = 512;
  #[cfg(feature = "large_data_set")]
  const HEIGHT : usize = 512;

  // Input and output matrix data
  let mut mat_in : Vec<i32> = vec![0; WIDTH*HEIGHT];
  let mut mat_out : Vec<i32> = vec![0; WIDTH*HEIGHT];

  // Initialise inputs
  let mut seed : u32 = 1;
  for i in 0..HEIGHT {
    for j in 0..WIDTH {
      mat_in[i*WIDTH + j] = rand15(&mut seed) as i32
    }
  }

  // Number of loop iterations per block.  The number of iterations
  // times the block Y dimension must equal the block X dimension.
  const ITERS_PER_BLOCK : usize = 4;

  // Block/grid dimensions
  let dims =
    Dims {
      block_dim: Dim2 { x: SQUARE_SIZE,
                        y: SQUARE_SIZE / ITERS_PER_BLOCK },
      grid_dim: Dim2 { x: WIDTH / SQUARE_SIZE,
                       y: HEIGHT / SQUARE_SIZE }
    };
 
  // Kernel parameters
  let mut params =
    Transpose { width  : WIDTH,
                height : HEIGHT,
                input  : mat_in.into(),
                output : mat_out.into() };

  // Invoke kernel
  let params = nocl_run_kernel_verbose(dims, params);

  // Check result
  let mut ok = true;
  for i in 0 .. HEIGHT {
    for j in 0 .. WIDTH {
      ok = ok && params.output[j*HEIGHT + i] ==
                 params.input[i*WIDTH + j];
    }
  }

  // Display result
  putstr("Self test: ");
  putstr(if ok { "PASSED" } else { "FAILED" });
  putchar(b'\n');

  // The end
  end()
}

File: apps/Samples/Transpose/Transpose.cpp
#include <NoCL.h>
#include <Rand.h>

// Kernel for matrix transposition
// One sub-square at a time
template <int SquareSize> struct Transpose : Kernel {
  Array2D<int> in, out;

  // Shared local memory
  Array2D<int> square;

  INLINE void init() {
    declareShared(&square, SquareSize, SquareSize+1);
  }
 
  INLINE void kernel() {
    // Origin of square within matrix
    int originX = blockIdx.x * blockDim.x;
    int originY = blockIdx.y * blockDim.x;
    
    // Load square
    for (int y = threadIdx.y; y < blockDim.x; y += blockDim.y)
      square[y][threadIdx.x] = in[originY + y][originX + threadIdx.x];
    
    __syncthreads();
    
    // Store square
    for (int y = threadIdx.y; y < blockDim.x; y += blockDim.y)
      out[originX + y][originY + threadIdx.x] = square[threadIdx.x][y];
  }
};

int main()
{
  // Are we in simulation?
  bool isSim = getchar();

  // Matrix size for benchmarking
  int width = isSim ? 256 : 512;
  int height = isSim ? 64 : 512;

  // Input and output matrix data
  nocl_aligned int matInData[width*height];
  nocl_aligned int matOutData[width*height];

  // Friendly array wrappers
  Array2D<int> matIn(matInData, height, width);
  Array2D<int> matOut(matOutData, width, height);

  // Initialise inputs
  uint32_t seed = 1;
  for (int i = 0; i < height; i++)
    for (int j = 0; j < width; j++)
      matIn[i][j] = rand15(&seed);

  // Number of loop iterations per block.  The number of iterations
  // times the block Y dimension must equal the block X dimension.
  const int itersPerBlock = 4;

  // Instantiate kernel
  Transpose<SIMTLanes> k;

  // Set block/grid dimensions
  k.blockDim.x = SIMTLanes;
  k.blockDim.y = SIMTLanes / itersPerBlock;
  k.gridDim.x = width / k.blockDim.x;
  k.gridDim.y = height / (itersPerBlock * k.blockDim.y);

  // Assign parameters
  k.in = matIn;
  k.out = matOut;

  // Invoke kernel
  noclRunKernelAndDumpStats(&k);

  // Check result
  bool ok = true;
  for (int i = 0; i < width; i++)
    for (int j = 0; j < height; j++)
      ok = ok && matOut[i][j] == matIn[j][i];

  // Display result
  puts("Self test: ");
  puts(ok ? "PASSED" : "FAILED");
  putchar('\n');

  return 0;
}
